compiler/noirc_frontend/src/elaborator/expressions.rs

```diff
@@ -846,19 +946,44 @@ impl<'context> Elaborator<'context> {
         (HirExpression::Tuple(element_ids), Type::Tuple(element_types))
     }
 
-    fn elaborate_lambda(&mut self, lambda: Lambda) -> (HirExpression, Type) {
+    /// For elaborating a lambda we might get `parameters_type_hints`. These come from a potential
+    /// call that has this lambda as the argument.
+    /// The parameter type hints will be the types of the function type corresponding to the lambda argument.
+    fn elaborate_lambda(
+        &mut self,
+        lambda: Lambda,
+        parameters_type_hints: Option<&Vec<Type>>,
+    ) -> (HirExpression, Type) {
         self.push_scope();
         let scope_index = self.scopes.current_scope_index();
 
         self.lambda_stack.push(LambdaContext { captures: Vec::new(), scope_index });
 
         let mut arg_types = Vec::with_capacity(lambda.parameters.len());
-        let parameters = vecmap(lambda.parameters, |(pattern, typ)| {
-            let parameter = DefinitionKind::Local(None);
-            let typ = self.resolve_inferred_type(typ);
-            arg_types.push(typ.clone());
-            (self.elaborate_pattern(pattern, typ.clone(), parameter, true), typ)
-        });
+        let parameters: Vec<_> = lambda
+            .parameters
+            .into_iter()
+            .enumerate()
+            .map(|(index, (pattern, typ))| {
+                let parameter = DefinitionKind::Local(None);
+                let is_unspecified = matches!(typ.typ, UnresolvedTypeData::Unspecified);
+                let typ = self.resolve_inferred_type(typ);
+
+                if is_unspecified {
```

Comment:
Shouldn't need this check here, we can resolve the inferred type which will return a type variable if it was unspecified. Then we can just unify the result unconditionally.